### PR TITLE
Added ability to specify a version range when installing snap

### PR DESCRIPTION
**Pull request by ritave**
fixes: #222

### CODE REVIEW

File: packages/controllers/src/snaps/SnapController.ts [hunk 1]
```diff
@@ -1415,9 +1429,8 @@ export class SnapController extends BaseController<
   }
 }
 
-function isValidSnapVersion(version: unknown): version is string {
+function isValidSnapVersionRange(version: unknown): version is string {
   return Boolean(
-    typeof version === 'string' &&
-      (version === 'latest' || isValidSemver(version)),
+    typeof version === 'string' && semver.validRange(version) !== null,
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/SnapController.ts:1434]: Is it intentional to remove the option to use `version === 'latest'`

Comment (ritave) [reply packages/controllers/src/snaps/SnapController.ts:1434]: Yeah, latest was removed from all places where semver is expected. This is because it's behaviour is not clearly defined.
For example, what should happen when a dapp requests latest while a user has one already installed? In case of npm, that would mean fetching config file and checking it again. In case of local development - well, there's no log of manifests, so what would "latest" mean in that scenario?

Additionally it would break all existing snaps and examples on every update in npm

Comment (FrederikBolding) [reply packages/controllers/src/snaps/SnapController.ts:1434]: 👍

File: packages/controllers/src/snaps/utils.ts [hunk 2]
```diff
@@ -321,10 +323,16 @@ async function fetchNpmTarball(
     );
   }
 
-  const targetVersion =
-    version === 'latest'
-      ? (packageMetadata as any)['dist-tags']?.latest
-      : version;
+  const targetVersion = semver.maxSatisfying(
+    Object.keys((packageMetadata as any)?.versions || {}),
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/utils.ts:327]: Would this be a good place to use `??` instead?

File: packages/controllers/src/snaps/utils.ts [hunk 3]
```diff
@@ -26,9 +27,11 @@ export enum NpmSnapFileNames {
   Manifest = 'snap.manifest.json',
 }
 
-export const LOCALHOST_HOSTNAMES = new Set(['localhost', '127.0.0.1']);
+export const LOCALHOST_HOSTNAMES = new Set(['localhost', '127.0.0.1', '::1']);
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/utils.ts:30]: Why is this needed?

File: packages/controllers/src/snaps/utils.ts (hunk 2 above)
Comment (FrederikBolding) [reply packages/controllers/src/snaps/utils.ts:327]: Also, why is the cast needed here?

File: packages/controllers/src/snaps/SnapController.ts [hunk 4]
```diff
@@ -14,7 +14,7 @@ import { SerializedEthereumRpcError } from 'eth-rpc-errors/dist/classes';
 import type { Patch } from 'immer';
 import { Json } from 'json-rpc-engine';
 import { nanoid } from 'nanoid';
-import isValidSemver from 'semver/functions/valid';
+import semver from 'semver';
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/SnapController.ts:17]: Would we get a bundle size benefit by importing only the used `semver` functions or can they be tree-shaken when importing everything?

**Review by FrederikBolding — COMMENTED** [review]
Overall this looks good, just left a few questions!

**Comment by FrederikBolding** [comment]
Also please rebase! 😄

**Review by FrederikBolding — APPROVED** [review]
